dbms/src/Common/MemoryTracker.cpp

```diff
@@ -93,20 +95,33 @@ void MemoryTracker::alloc(Int64 size, bool check_memory_limit)
 
             throw DB::TiFlashException(fmt_buf.toString(), DB::Errors::Coprocessor::MemoryLimitExceeded);
         }
-
-        if (unlikely(current_limit && will_be > current_limit))
+        bool is_rss_too_large = (current_limit
+                                 && real_rss > current_limit + 5LL * 1024 * 1024 * 1024
+                                 && will_be > current_limit - (real_rss - current_limit - 5LL * 1024 * 1024 * 1024));
```

Comment:
since real_rss metric  is updated every 100ms, if we only check “real_rss > current_limit + 5LL * 1024 * 1024 * 1024”，it cause all mem alloc failed in current 100ms period. So we can define a linear check logic. A example for how this logic works: if `limit` is 30GB, and `real_rss` is 40GB, since then mem alloc will fail when `tracked amount` > 25GB(30-(10-5)=25), when memory usage turn down below 25GB, all query tasks can continue work, where there is no need to wait the `real_rss` metric updated in next collect 100ms period. The larger real_rss over than limit+5GB, the lower memory tracked `amout` ceiling we can continue allocating memory.

Reply:
sometimes `current_limit` is not a global limit, for example, if only per query's limit is defined, then it seems not work well?
rss = 100G
current_limit = 20G
then will_be > (20G - (100G - 20G - 5G)) => will_be > -55G, which is always true.

Reply:
good spot, `current_limit` should by global limit , it was a mistake

Reply:
fixed

Reply:
good idea